[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61942
* #61892
* #61859
* #61763
* __->__ #61687

Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

jerryzh168 committed:
[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes

Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

vkuzo commented:
just curious, what's the reasoning for doing this for all cases compared to only cases where is_reference is true

jerryzh168 commented:
> just curious, what's the reasoning for doing this for all cases compared to only cases where is_reference is true

This change is valid for both cases I think, we are not modeling copy nodes correctly before in qat, we should have an observer that's shared by both input and output since quantized copy node (e.g. quantized maxpool) assumes that (we don't provide scale, zero_point for output of maxpool since it's inherited from the input).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

vkuzo commented:
makes sense.  What specifically is bc breaking?  Would models prepared with code "before" not be able to converted with code "after", or something else?

jerryzh168 commented:
> makes sense. What specifically is bc breaking? Would models prepared with code "before" not be able to converted with code "after", or something else?

Updated BC-breaking notes in the summary, please take a look

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -548,9 +548,10 @@ def maybe_insert_output_observer_for_node(
     should_insert_observer = \
         qhandler.should_insert_observer_for_output(
-            qconfig, model.training)
+            qconfig, model.training) and dtype not in (torch.bool, None, torch.float)
```
should we add dtype as an argument to `should_insert_observer_for_output`, so that the logic is in one place?

jerryzh168 replied:
yeah, that sounds good, will add in a follow up PR

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).

jerryzh168 committed:
Update on "[bc-breaking][quant][graphmode][fx] Add observer/fake_quant for copy nodes"


Summary:
Previously we do not insert observer/fake_quant for output copy nodes (e.g. maxpool).
But to produce reference patterns we need to insert observer/fake_quant for the output and later convert that to a quantize
node.

BC-breaking notes:
This PR changes the numerics of QAT for nodes like maxpool, previously we only insert one fake quant before maxpool, after this PR, we would also insert another fake quant instance to the output of maxpool, this change makes the conversion step exact for the qat model with these ops.
Before: fake_quant - maxpool - ...
After: fake_quant - maxpool - fake_quant (same fake_quant instance as input) - ...



Model:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.maxpool2d = torch.nn.MaxPool2d(kernel_size=3)

    def forward(self, x):
        x = self.maxpool2d(x)
        return x
```
result of prepare:

Before:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    return maxpool2d

After:
def forward(self, x):
    x_activation_post_process_0 = self.x_activation_post_process_0(x);  x = None
    maxpool2d = self.maxpool2d(x_activation_post_process_0);  x_activation_post_process_0 = None
    maxpool2d_activation_post_process_0 = self.maxpool2d_activation_post_process_0(maxpool2d);  maxpool2d = None
    return maxpool2d_activation_post_process_0

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29715566](https://our.internmc.facebook.com/intern/diff/D29715566)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29715566).
